presentation/frontend/src/presentation-frontend/StateTracker.ts

```diff
@@ -47,77 +47,68 @@ export class StateTracker {
+      expandedNode.expandedIn.delete(sourceId);
+      // if there are other sources that have this node expanded leave it.
+      if (expandedNode.expandedIn.size !== 0)
         continue;
 
-      hierarchy.delete(collapsedNode.id);
-      removedKeys.push(collapsedNode.key);
+      hierarchy.delete(nodeId);
+      removedKeys.push(expandedNode.key);
     }
 
-    // if hierarchy does not have any nodes remove it
     if (hierarchy.size === 0)
       this._expandedHierarchies.delete(rulesetId);
 
     await this.updateHierarchyStateIfNeeded(imodel.key, rulesetId, "nodesCollapsed", removedKeys);
   }
 
-  public async onHierarchyClosed(imodel: IModelConnection, rulesetId: string, sourceId: string) {
-    const hierarchy = this._expandedHierarchies.get(rulesetId);
-    if (!hierarchy)
+  public async onExpandedNodesChanged(imodel: IModelConnection, rulesetId: string, sourceId: string, expandedNodes: NodeIdentifier[]) {
+    let hierarchy = this._expandedHierarchies.get(rulesetId);
+    if (expandedNodes.length === 0 && !hierarchy)
       return;
+    if (!hierarchy) {
+      hierarchy = new Map<string, ExpandedNode>();
+      this._expandedHierarchies.set(rulesetId, hierarchy);
+    }
 
     const removedKeys: NodeKey[] = [];
-    for (const [nodeId, expandedNode] of hierarchy) {
-      expandedNode.expandedIn.delete(sourceId);
-      // if there are other sources that have this node expanded leave it.
-      if (expandedNode.expandedIn.size !== 0)
+    const addedKeys: NodeKey[] = [];
+    for (const [key, existingNode] of hierarchy) {
+      // existing node is in new expanded nodes list. Add current source
+      if (expandedNodes.find((expandedNode) => expandedNode.id === key)) {
+        existingNode.expandedIn.add(sourceId);
         continue;
+      }
 
-      hierarchy.delete(nodeId);
-      removedKeys.push(expandedNode.key);
+      // node was not found in expanded nodes list. Remove current source
+      existingNode.expandedIn.delete(sourceId);
+      if (existingNode.expandedIn.size !== 0)
+        continue;
+
+      removedKeys.push(existingNode.key);
+      hierarchy.delete(key);
     }
 
-    if (hierarchy.size === 0)
-      this._expandedHierarchies.delete(rulesetId);
+    // add any new nodes that were not in expanded nodes hierarchy already
+    for (const expandedNode of expandedNodes) {
+      const existingNode = hierarchy.get(expandedNode.id);
+      if (existingNode)
+        continue;
 
-    await this.updateHierarchyStateIfNeeded(imodel.key, rulesetId, "nodesCollapsed", removedKeys);
+      hierarchy.set(expandedNode.id, { key: expandedNode.key, expandedIn: new Set<string>([sourceId]) });
+      addedKeys.push(expandedNode.key);
+    }
+
+    if (removedKeys.length !== 0)
+      await this.updateHierarchyStateIfNeeded(imodel.key, rulesetId, "nodesCollapsed", removedKeys);
+    if (addedKeys.length !== 0)
+      await this.updateHierarchyStateIfNeeded(imodel.key, rulesetId, "nodesExpanded", addedKeys);

```

Comment:
- The `updateHierarchyStateIfNeeded` does the empty keys list check internally, so no need to that here.
- These two calls will result in two IPC requests. Maybe not a big deal, but look like this approach would be better:
  ```ts
  this.updateHierarchyStateIfNeeded(imodel.key, rulesetId, { nodesExpanded: addedKeys, nodesRemoved: removedKeys });
  ```

Reply:
Removed checks for empty lists. I don't think that when expanded nodes change there should be a lot of cases where added and removed keys would be found. Usually it will be called when node is expanded/collapsed in the tree or nodes are added/removed to/from model. So it should not result in two IPC request always.

As for function signature taking `changeType` and `nodeKeys` array I was hoping to reuse this for tracking visible nodes. It should be enough to add new `changeType` without changing function signature or adding new IPC function